### ISSUE TITLE
Covariance error inferred for lambda has invalid line number associated with it

### ISSUE THREAD

**Issue by kprzybyla** [python/mypy#8592]
#### Versions

`mypy 0.770`
`python 3.8.2`

#### Description

This is probably related to the https://github.com/python/mypy/issues/4270

Problem is triggered when using recipe `everything_except` from hypothesis documentation:

https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.from_type

I have a file with following content (this is the whole file):

```python
from typing import TypeVar

from hypothesis import strategies as st

_Ex = TypeVar("_Ex", covariant=True)


def everything_except(*types: type) -> st.SearchStrategy[_Ex]:
    strategy = st.from_type(type).flatmap(st.from_type)

    return strategy.filter(lambda x: not isinstance(x, types))  # type: ignore
```

Running mypy on first clean run (without `.mypy_cache`) has the following output:

```
example.py:1276: error: Cannot use a covariant type variable as a parameter
Found 1 error in 1 file (checked 1 source file)
```

For some reason, `mypy` sees error on line `1276` which does not exist. On the other hand, running `mypy` for the second time gives:

```
Success: no issues found in 1 source file
```

What's more, when you have `pretty` set to `True`, first run gives:

```
Traceback (most recent call last):
  File "/home/krprzyby/.local/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  File "/home/krprzyby/.local/lib/python3.8/site-packages/mypy/__main__.py", line 8, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "mypy/main.py", line 89, in main
  File "mypy/build.py", line 180, in build
  File "mypy/build.py", line 249, in _build
  File "mypy/build.py", line 2649, in dispatch
  File "mypy/build.py", line 2956, in process_graph
  File "mypy/build.py", line 3073, in process_stale_scc
  File "mypy/errors.py", line 476, in file_messages
  File "mypy/errors.py", line 449, in format_messages
IndexError: list index out of range
```

Running `mypy` for the second time works correctly with `pretty` set to `True` because the line number is correct and `# type: ignore` works.

**Comment by kprzybyla** [comment]
Also, for anyone that faces similar issue when using this recipe from `hypothesis` documentation, here's a working solution that avoids any `mypy` errors and do not require `# type: ignore`

```python
from typing import TypeVar

from hypothesis import strategies as st

_T = TypeVar("_T")
_Ex = TypeVar("_Ex", covariant=True)

def everything_except(*types: type) -> st.SearchStrategy[_Ex]:
    def filter_everything_except(value: _T) -> bool:
        return not isinstance(value, types)

    return st.from_type(type).flatmap(st.from_type).filter(filter_everything_except)
```

**Comment by ilevkivskyi** [comment]
Wrong line number for error is a pretty serious issue, hard to say what is a possible root cause however.

**Comment by pranavrajpal** [comment]
I think the problem is that the check for whether a covariant type variable is used as the argument to a function reports the line number that the type variable is bound to a variable (which makes sense when that type variable is an argument for a `FuncDef` but doesn't make sense for a `Callable` variable or a lambda).

The wrong line number (which was line 884 in my case) seems to be pointing [here](https://github.com/HypothesisWorks/hypothesis/blob/f8932c4f38f91c882a4cce75f86a572b976e9ac4/hypothesis-python/src/hypothesis/strategies/_internal/core.py#L884).

I managed to create a somewhat simpler example:
```py
from typing import Any, Callable, TypeVar

Ex = TypeVar("Ex")
T = TypeVar("T", covariant=True)

def takes_lambda(func: Callable[[Ex], Any]) -> None:
    pass

def test(a: T) -> None: # E: Cannot use a covariant type variable as a parameter
    func: Callable[[T], Any]
    takes_lambda(func) # type: ignore
```
